synapse/types.py

```diff
@@ -421,22 +421,92 @@ class RoomStreamToken:
 
             s0    s1
             |     |
-        [0] V [1] V [2]
+        [0] ▼ [1] ▼ [2]
 
     Tokens can either be a point in the live event stream or a cursor going
     through historic events.
 
-    When traversing the live event stream events are ordered by when they
+    When traversing the live event stream, events are ordered by when they
     arrived at the homeserver.
 
-    When traversing historic events the events are ordered by their depth in
-    the event graph "topological_ordering" and then by when they arrived at the
-    homeserver "stream_ordering".
+    When traversing historic events, events are first ordered by their
+    "depth" ("topological_ordering" in the event graph) and tie-broken by
+    "stream_ordering" (when the event arrived at the homeserver).
+
+    ---
 
     Live tokens start with an "s" followed by the "stream_ordering" id of the
-    event it comes after. Historic tokens start with a "t" followed by the
-    "topological_ordering" id of the event it comes after, followed by "-",
-    followed by the "stream_ordering" id of the event it comes after.
+    event that comes after. The rest of the keys are joined together by
```

Comment:
Does the token represent a position before or after the event with the given stream_ordering? I'm confused as to which is right.

Reply:
I don't know for sure but based on the previous comments, I would say the token represents the position before the given "id". But the "id" itself is the one that comes after the position. Can be phrased either way. Better way to explain?

Here are the previous comments around this.

> Live tokens start with an "s" followed by the "stream_ordering" id of the event it comes after.
    
> Historic tokens start with a "t" followed by the "topological_ordering" id of the event it comes after

Reply:
After a bit of a call with @erikjohnston:
`stream_ordering`s uniquely identify persisted events. When a `RoomStreamToken` contains a `stream_ordering`, the token means "the position just after the event identified by `stream_ordering`". I'm not sure if it works differently for backfill.

That is, I think the wording should be "the event it comes after" instead of "the event that comes after".
And the same for line +491.

Reply:
The nuance here is making me read this 1000 times 😵

---

The first explanation makes sense 👍

> `stream_ordering`s uniquely identify persisted events. When a `RoomStreamToken` contains a `stream_ordering`, the token means "the position just after the event identified by `stream_ordering`".


How does this sound?

```diff
+ Live tokens start with an "s" followed by the "stream_ordering" id of the
+ event that comes before position of the token.
```